README.md
[![](https://vsmarketplacebadge.apphb.com/version-short/niudai.vscode-zhihu.svg)](https://marketplace.visualstudio.com/items?itemName=niudai.vscode-zhihu)
[![](https://vsmarketplacebadge.apphb.com/downloads-short/niudai.vscode-zhihu.svg)](https://marketplace.visualstudio.com/items?itemName=niudai.vscode-zhihu)
[![](https://vsmarketplacebadge.apphb.com/rating-short/niudai.vscode-zhihu.svg)](https://marketplace.visualstudio.com/items?itemName=niudai.vscode-zhihu)


<p align="center">
  <img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/vscode-zhihu.png" alt="vscode-zhihu logo" width="200px" /></a>
</p>
 
<p align="center">
<a href="https://github.com/niudai/VSCode-Zhihu">打一颗 ⭐，世界更亮。</a>
</p> 


# 👽 Zhihu On VSCode

基于 VSCode 的知乎客户端提供包括阅读，搜索，创作，发布等一站式服务，内容加载速度比 Web 端更快，创新的 Markdown-Latex 混合语法让内容创作者更方便地插入代码块，数学公式，并一键发布至知乎平台。项目由 [牛岱](https://www.zhihu.com/people/niu-dai-68-44) 独立设计开发，喜欢的话请献出你的 [⭐](https://github.com/niudai/VSCode-Zhihu '给一个Star')。

## ⚡ Features

- 登录
  - [二维码/账密登录](#🔑-登录 )
- 创作
  - [内容创作](#🖍-内容创作)
  - [内容发布](#📩-内容发布)
  - [一键上传图片](#📊-上传图片)
  - [定时发布](#🕐-定时发布)
- 浏览  
  - [个性推荐](#🎭-个性推荐)
  - [实时热榜](#hot-story)
  - [搜索全站](#🔎-搜索 )
  - [收藏夹](#🎫-收藏夹)


## 📃 Reference

- [图标按钮](#😀-图标按钮)
- [快捷键](#⌨-快捷键)
- [配置项](#⚙-配置项)

---

## 🔑 登录 


<a name = "login"></a>

进入主页面，左侧最上方栏为个人中心，点击登录图标，或使用 `Ctrl + Shift + P` 打开命令面板，搜索并执行 `Zhihu: Login` 命令。

选择登录方式：

### 二维码

选择二维码登陆后，会弹出二维码页面，打开知乎 APP，扫码后点击确认登录：

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-08-20-28-08.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

### 账号密码

视情况，插件会加载并显示验证码，提示你输入验证码，输入后，再依次根据提示输入手机号和密码即可。

登录成功后会有问候语，推荐栏会自动刷新出你的个性签名和头像：

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-08-20-30-17.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-08-20-32-09.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

    

---

## 🎭 个性推荐

登陆成功后，个性推荐板块会自动刷新，提供你的个性推荐内容：

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-08-20-30-27.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

内容可能为答案，问题，或文章，点击条目，就会打开VSCode知乎页面：

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-08-21-02-30.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

___

## 💥 热榜

在左侧的中间位置，你会看到热榜栏，内部有六个分类，内容与知乎Web端、移动端同步，助你掌控实时资讯：

<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-10-11-01-37.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

---

## 🔎 搜索 

<a name = "search"></a>
点击搜索按钮，或搜索命令 `Zhihu: Search Items`，搜索全站知乎内容：

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/search.gif" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

---

## 🖍 内容创作
<a name = "write"></a>

新建一个后缀名为`.md`的文件，若不需要数学公式，只需要按照你最熟悉的 Markdown 语法写即可，右键点击 `Zhihu: Preview` 可预览答案:

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/preview.gif" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

### Latex 语法支持

为了更好地支持数学公式的写作，知乎定制的 Markdown 转换器提供了 Latex 语法拓展，语法示例：

```
$$
  |\vec{A}|=\sqrt{A_x^2 + A_y^2 + A_z^2}.
$$
```

用 `$$` 包围的部分会被当做 latex 语言进行解析，生成知乎的数学公式，比如上方的数学公式发布至知乎会生成如下公式:

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-08-22-42-21.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

行内 latex 也同样支持，语法举例：`$\sqrt6$`，一个dollar符号包裹公式即可。

代码块：

记得声明语言标签, 这样发布至知乎的答案才能获得正确的语法高亮，示例如下：

    ```java
    public class Apple {
      public Apple() {}
    }
    ```

发布后会提供 java 的语法高亮:

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-08-22-47-18.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

>由于知乎服务端的限制，表格暂不支持，答案中的表格会被服务端过滤。

## 📩 内容发布
<a name = "publish"></a>

### 链接扫描 😊

若你想在特定的问题下回答，或想修改自己的某个原有回答，就将问题/答案链接以 `#! https://...` 的格式放置于答案的第一行，发布时，插件会自动扫描识别，发布至相应的问题下，或修改原有的答案。

比如，你想在 [轻功是否真的存在，其在科学上可以解释吗？](https://www.zhihu.com/question/19602618) 该问题下回答问题, 只需将

```
#! https://www.zhihu.com/question/19602618
```

若是你已经创作过的答案, 则将答案的链接, 形如:

```
#! https://www.zhihu.com/question/355223335/answer/1003461264
```

的链接复制至文件顶部即可。

若是你已经创作过的文章，则将文章的链接，形如：

```
#! https://zhuanlan.zhihu.com/p/107810342
```

若插件没有在首行扫描到链接，则会询问创作者接下来的操作，你可以选择发布新文章，或从收藏夹中选取相应问题，发布至相应问题下：

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-08-20-47-45.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

### 发布文章

选择发布文章后，会继续提示你输入文章标题，输入完成后，按下回车，当前的文档就会以文章的形式发布至你的账号。

#### 文章标题智能识别

文章标题无需手动输入，插件会自动检测文本的第一个一级头标签：

```
# 这是一个标题（必须只是一个#）
```

然后将其作为标题，改行的内容也不会进入到正文中，如果没有检测到，还需用户手动输入。

#### 背景图片智能识别

插件会自动扫描文本第一个一级头标签之前的内容，将第一个发现的图片链接作为背景图片：

```
![Image](https://pic4.zhimg.com/80/v2-157583e100e9e181191d285355332ebf.png)

# 标题在这, 上面的链接会变成背景图片, 不会进入正文
```

### Html 支持

可以在正常的 Markdown 文本中插入 html 文本, 扩展了写作能力。

>绝大多数 html 标签为非法标签，包括 table 在内，会被服务端过滤掉，只有 \<p\>, \<div>, \<img> 等合法标签才会被服务端存储，具体使用时小伙伴们可以自己尝试。

### 从收藏夹中选取

>关于如何管理收藏夹，请移至 [收藏夹](#collect)。

插件会提示选择你收藏过的问题：

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-08-20-51-43.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

选择后，答案就会发布至相应的答案下（若已在该答案下发布过问题，请用顶部链接的方式来发布！)。

---

## 🕐 定时发布

所有的答案，文章发布时，均会多一次询问，用户须选择是稍后发布还是马上发布，如果选择稍后发布，需要输入发布的时间，比如 “5:30 pm”，"9:45 am" 等，目前仅支持当天的时间选择，输入后，你就会在个人中心的“安排”处看到你将发布的答案和发布的时间（需要手动点击刷新）：

![](https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-16-11-20-14.png)

定时发布采用 prelog 技术，中途关闭 VSCode，关机不影响定时发布，只需保证发布时间 VSCode 处于打开状态 && 知乎插件激活状态即可。

时间到了之后，你会收到答案发布的通知，该事件也会从“安排”中移除。

如果想取消发布，则点击 ❌ 按钮即可：

![](https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-16-15-56-31.png)

>发布事件采用 md5 完整性校验，不允许用户同时预发两篇内容一摸一样的答案或文章。
---

## 🎫 收藏夹
<a name = "collect"></a>

### ➕ 添加收藏

不管是文章，答案，还是问题，在知乎页面顶栏的右侧，都会看到一个粉色的星状图标：

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-08-20-56-42.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

### ➖ 查看收藏

收藏的内容会在左侧下方显示，插件会自动分类：

<p align="center">
<img src="https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-08-20-57-40.png" style="box-shadow: 2px 2px 8px 0px #5dd8fd;border-radius: 6px;"/></p>

### ✖ 删除收藏

鼠标移至相应的行，会出现叉状图标，点击即可删除：

<p align="center">

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.0]

### 文章/答案发布后自动生成头部链接

在以往的版本中，发布一篇新文章，新答案，就会产生新的链接，但是如果想修改这篇文章或答案，就需要用 `#! https://zhuanlan.zhihu.com/p/126167760` 形式的链接置于文章顶部，在 0.3 版本中，该操作插件会替你自动完成，也就是说，一份源 markdown 文件，发布后会自动指向对应的文章和答案，修改后再发布，即可修改源文章。

### 域外图片缓存加速

上一个版本中，插件支持了域外链接和本地链接的图片，但是使用起来会发现，发布的时间比较长，尤其是有域外链接的图片时，插件会先把域外的图片下载到本地，再传到知乎图床上，完成链接的替换，在 0.3 版本中，已经上传过的域外链接，插件会有缓存记录，再次发布时，会直接完成链接替换，跳过下载和上传过程。

如要清理缓存，请使用 `zhihu.clearCache` 命令。

### 支持本地绝对路径

上一个版本，图片链接只支持本地相对路径和域外 https 链接，新版本支持本地的绝对路径。（请注意 http:// 协议仍然不支持，请保证域外图片为 https://)

## [0.2.3]

众所周知，一个来自外域的图片链接，是不一定能在知乎平台上正常显示的，因为会涉及到跨域问题，为了安全起见，原则上所有答案/文章中的图片，都要上传至知乎的图床，然后将链接放在答案中，这样才能正常显示。

不仅如此，知乎服务端也不允许上传的答案或文章中的图片的来源为非知乎的源，即便我们试图这样做，其实也是不可行的。

所以在图片链接上这点，用户需要用插件自身提供的图片上传功能，上传至知乎图床，插件也会自动在 Markdown 文本里插入图片链接，其实已经比较方便了，但不能算作“非侵入式”，因为这改变了用户本来的习惯。

于是小岱开始开发了一个新 Feature，这个 Feature 可以让用户无需在意图片链接是否是来自知乎图床，这个链接可以是本地的相对路径，也可以来自于知乎域外，发布至知乎时，所有的图片都能够正常显示。

也就是说，随便在你的电脑中拿出一个你以前写好的某个 README，就算这个 README 里面有一堆相对路径的图片，或来自奇奇怪怪的图床的图片，只需右键点击发布，it just works。

比如：

![Image](https://pic4.zhimg.com/80/v2-0b00790259520bdbda398cd05731b06b.png)

源 Markdown 中的所有图片都是相对路径或来自外链。

发布后的效果：

![Image](https://pic4.zhimg.com/80/v2-22d902f8c869bc61c44c7711fa8e4e00.png)


## [0.2.2]

### 专栏管理

为了让创作者群体更好地创作，在发布文章的时候，用户可以直接选择发布至自己的专栏下：

![Image](https://pic4.zhimg.com/80/v2-b6358e6d673e5feb84dd0ad0bd4d52e4.png)

### 文章标题智能识别

文章标题无需手动输入，插件会自动检测文本的第一个一级头标签：

```
# 这是一个标题（必须只是一个#）
```

然后将其作为标题，改行的内容也不会进入到正文中，如果没有检测到，还需用户手动输入。

### 背景图片智能识别

插件会自动扫描文本第一个一级头标签之前的内容，将第一个发现的图片链接作为背景图片：

```
![Image](https://pic4.zhimg.com/80/v2-157583e100e9e181191d285355332ebf.png)
```
```
# 标题在这, 上面的链接会变成背景图片, 不会进入正文
```

### Html 支持

可以在正常的 Markdown 文本中插入 html 文本, 扩展了写作能力。

>绝大多数 html 标签为非法标签，包括 table 在内，会被服务端过滤掉，只有 \<p\>, \<div>, \<img> 等合法标签才会被服务端存储，具体使用时小伙伴们可以自己尝试。

### 增加 Zhihu: Is Title Image Full Screen 配置项

用户可以在设置中找到 `Zhihu: Is Title Image Full Screen` 配置项，勾选后，知乎文章的背景图片会变为全屏模式。

## [0.2.1]

### 可以查看点赞数，并给喜欢的内容点赞

![Image](https://pic4.zhimg.com/80/v2-d8f61703c731711fe3a3585122c0d676_hd.png)

点击按钮点赞。

### 显示作者头像，名字，个性签名

![Image](https://pic4.zhimg.com/80/v2-157583e100e9e181191d285355332ebf_hd.png)

点击头像，可以在浏览器打开该作者的个人主页。

### 修改文章

发布后的文章，按照和答案相同的方式，将文章的链接以形如：

```
#! https://zhuanlan.zhihu.com/p/107810342
```

复制至文章顶部，发布即可对原文章进行修改。

### 行内latex

现在创作的时候，可以直接用 `$\sqrt5$` 的方式写行内 latex, 而块latex还是原来的 `$$\sqrt$$` 语法。

### 优化了图标样式

![Image](https://pic4.zhimg.com/80/v2-b293132e3d47d8cd48394caf78611bd2_hd.png)

## [0.2.0] - 2020-02-17

### Webview 默认使用 VSCode 主题色

板块是透明的，会看起来像透明亚克力：

![](https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-16-11-11-22.png)

>可以在 VSCode 的设置栏中找到 `Use VSTheme` 设置项，取消打勾后，会开启知乎默认的白蓝主题。

### 支持定时发布

所有的答案，文章发布时，均会多一次询问，用户须选择是稍后发布还是马上发布，如果选择稍后发布，需要输入发布的时间，比如 “5:30 pm”，"9:45 am" 等，目前仅支持当天的时间选择，输入后，你就会在个人中心的“安排”处看到你将发布的答案和发布的时间（需要手动点击刷新）：

![](https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-16-11-20-14.png)

定时发布采用 prelog 技术，中途关闭 VSCode，关机不影响定时发布，只需保证发布时间 VSCode 处于打开状态 && 知乎插件激活状态即可。

时间到了之后，你会收到答案发布的通知，该事件也会从“安排”中移除。

如果想取消发布，则点击 ❌ 按钮即可：

![](https://raw.githubusercontent.com/niudai/ImageHost/master/zhihu/2020-02-16-15-56-31.png)


login.js
var fs = require('fs');
var httpClient = require('request-promise');
var path = require('path');
var crypto = require('crypto');
var formurlencoded = require('form-urlencoded').default;
var stdin = process.stdin;


//////////////////// Encryption Code //////////

function s(e) {
    return (s = "function" == typeof Symbol && "symbol" == typeof Symbol.t ? function(e) {
        return typeof e
    }
    : function(e) {
        return e && "function" == typeof Symbol && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : typeof e
    }
    )(e)
}
function i() {}
function h(e) {
    this.s = (2048 & e) >> 11,
    this.i = (1536 & e) >> 9,
    this.h = 511 & e,
    this.A = 511 & e
}
function A(e) {
    this.i = (3072 & e) >> 10,
    this.A = 1023 & e
}
function n(e) {
    this.n = (3072 & e) >> 10,
    this.e = (768 & e) >> 8,
    this.a = (192 & e) >> 6,
    this.s = 63 & e
}
function e(e) {
    this.i = e >> 10 & 3,
    this.h = 1023 & e
}
function a() {}
function c(e) {
    this.n = (3072 & e) >> 10,
    this.e = (768 & e) >> 8,
    this.a = (192 & e) >> 6,
    this.s = 63 & e
}
function o(e) {
    this.A = (4095 & e) >> 2,
    this.s = 3 & e
}
function r(e) {
    this.i = e >> 10 & 3,
    this.h = e >> 2 & 255,
    this.s = 3 & e
}
function k(e) {
    this.s = (4095 & e) >> 10,
    this.i = (1023 & e) >> 8,
    this.h = 1023 & e,
    this.A = 63 & e
}
function B(e) {
    this.s = (4095 & e) >> 10,
    this.n = (1023 & e) >> 8,
    this.e = (255 & e) >> 6
}
function f(e) {
    this.i = (3072 & e) >> 10,
    this.A = 1023 & e
}
function u(e) {
    this.A = 4095 & e
}
function C(e) {
    this.i = (3072 & e) >> 10
}
function b(e) {
    this.A = 4095 & e
}
function g(e) {
    this.s = (3840 & e) >> 8,
    this.i = (192 & e) >> 6,
    this.h = 63 & e
}
function G() {
    this.c = [0, 0, 0, 0],
    this.o = 0,
    this.r = [],
    this.k = [],
    this.B = [],
    this.f = [],
    this.u = [],
    this.C = !1,
    this.b = [],
    this.g = [],
    this.G = !1,
    this.Q = null,
    this.R = null,
    this.w = [],
    this.x = 0,
    this.D = {
        0: i,
        1: h,
        2: A,
        3: n,
        4: e,
        5: a,
        6: c,
        7: o,
        8: r,
        9: k,
        10: B,
        11: f,
        12: u,
        13: C,
        14: b,
        15: g
    }
}
Object.defineProperty(exports, "__esModule", {
    value: !0
});
var t = "1.1"
  , __g = {};
i.prototype.M = function(e) {
    e.G = !1
}
,
h.prototype.M = function(e) {
    switch (this.s) {
    case 0:
        e.c[this.i] = this.h;
        break;
    case 1:
        e.c[this.i] = e.k[this.A]
    }
}
,
A.prototype.M = function(e) {
    e.k[this.A] = e.c[this.i]
}
,
n.prototype.M = function(e) {
    switch (this.s) {
    case 0:
        e.c[this.n] = e.c[this.e] + e.c[this.a];
        break;
    case 1:
        e.c[this.n] = e.c[this.e] - e.c[this.a];
        break;
    case 2:
        e.c[this.n] = e.c[this.e] * e.c[this.a];
        break;
    case 3:
        e.c[this.n] = e.c[this.e] / e.c[this.a];
        break;
    case 4:
        e.c[this.n] = e.c[this.e] % e.c[this.a];
        break;
    case 5:
        e.c[this.n] = e.c[this.e] == e.c[this.a];
        break;
    case 6:
        e.c[this.n] = e.c[this.e] >= e.c[this.a];
        break;
    case 7:
        e.c[this.n] = e.c[this.e] || e.c[this.a];
        break;
    case 8:
        e.c[this.n] = e.c[this.e] && e.c[this.a];
        break;
    case 9:
        e.c[this.n] = e.c[this.e] !== e.c[this.a];
        break;
    case 10:
        e.c[this.n] = s(e.c[this.e]);
        break;
    case 11:
        e.c[this.n] = e.c[this.e]in e.c[this.a];
        break;
    case 12:
        e.c[this.n] = e.c[this.e] > e.c[this.a];
        break;
    case 13:
        e.c[this.n] = -e.c[this.e];
        break;
    case 14:
        e.c[this.n] = e.c[this.e] < e.c[this.a];
        break;
    case 15:
        e.c[this.n] = e.c[this.e] & e.c[this.a];
        break;
    case 16:
        e.c[this.n] = e.c[this.e] ^ e.c[this.a];
        break;
    case 17:
        e.c[this.n] = e.c[this.e] << e.c[this.a];
        break;
    case 18:
        e.c[this.n] = e.c[this.e] >>> e.c[this.a];
        break;
    case 19:
        e.c[this.n] = e.c[this.e] | e.c[this.a]
    }
}
,
e.prototype.M = function(e) {
    e.r.push(e.o),
    e.B.push(e.k),
    e.o = e.c[this.i],
    e.k = [];
    for (var t = 0; t < this.h; t++)
        e.k.unshift(e.f.pop());
    e.u.push(e.f),
    e.f = []
}
,
a.prototype.M = function(e) {
    e.o = e.r.pop(),
    e.k = e.B.pop(),
    e.f = e.u.pop()
}
,
c.prototype.M = function(e) {
    switch (this.s) {
    case 0:
        e.C = e.c[this.n] >= e.c[this.e];
        break;
    case 1:
        e.C = e.c[this.n] <= e.c[this.e];
        break;
    case 2:
        e.C = e.c[this.n] > e.c[this.e];
        break;
    case 3:
        e.C = e.c[this.n] < e.c[this.e];
        break;
    case 4:
        e.C = e.c[this.n] == e.c[this.e];
        break;
    case 5:
        e.C = e.c[this.n] != e.c[this.e];
        break;
    case 6:
        e.C = e.c[this.n];
        break;
    case 7:
        e.C = !e.c[this.n]
    }
}
,
o.prototype.M = function(e) {
    switch (this.s) {
    case 0:
        e.o = this.A;
        break;
    case 1:
        e.C && (e.o = this.A);
        break;
    case 2:
        e.C || (e.o = this.A);
        break;
    case 3:
        e.o = this.A,
        e.Q = null
    }
    e.C = !1
}
,
r.prototype.M = function(e) {
    switch (this.s) {
    case 0:
        for (var t = [], n = 0; n < this.h; n++)
            t.unshift(e.f.pop());
        e.c[3] = e.c[this.i](t[0], t[1]);
        break;
    case 1:
        for (var r = e.f.pop(), o = [], i = 0; i < this.h; i++)
            o.unshift(e.f.pop());
        e.c[3] = e.c[this.i][r](o[0], o[1]);
        break;
    case 2:
        for (var a = [], c = 0; c < this.h; c++)
            a.unshift(e.f.pop());
        e.c[3] = new e.c[this.i](a[0],a[1])
    }
}
,
k.prototype.M = function(e) {
    switch (this.s) {
    case 0:
        e.f.push(e.c[this.i]);
        break;
    case 1:
        e.f.push(this.h);
        break;
    case 2:
        e.f.push(e.k[this.A]);
        break;
    case 3:
        e.f.push(e.g[this.A])
    }
}
,
B.prototype.M = function(t) {
    switch (this.s) {
    case 0:
        var s = t.f.pop();
        t.c[this.n] = t.c[this.e][s];
        break;
    case 1:
        var i = t.f.pop()
          , h = t.f.pop();
        t.c[this.e][i] = h;
        break;
    case 2:
        var A = t.f.pop();
        if(A === 'window') {
            A = {
                encodeURIComponent: function (url) {
                    return encodeURIComponent(url)
                }
            }
        } else if (A === 'navigator') {
            A = {
                'userAgent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 ' +
                    '(KHTML, like Gecko) Chrome/66.0.3359.181 Safari/537.36'
            }
        }
        t.c[this.n] = eval(A)
    }
}
,
f.prototype.M = function(e) {
    e.c[this.i] = e.g[this.A]
}
,
u.prototype.M = function(e) {
    e.Q = this.A
}
,
C.prototype.M = function(e) {
    throw e.c[this.i]
}
,
b.prototype.M = function(e) {
    var t = this
      , n = [0];
    e.k.forEach(function(e) {
        n.push(e)
    });
    var r = function(r) {
        var o = new G;
        return o.k = n,
        o.k[0] = r,
        o.J(e.b, t.A, e.g, e.w),
        o.c[3]
    };
    r.toString = function() {
        return "() { [native code] }"
    }
    ,
    e.c[3] = r
}
,
g.prototype.M = function(e) {
    switch (this.s) {
    case 0:
        for (var t = {}, n = 0; n < this.h; n++) {
            var r = e.f.pop();
            t[e.f.pop()] = r
        }
        e.c[this.i] = t;
        break;
    case 1:
        for (var o = [], i = 0; i < this.h; i++)
            o.unshift(e.f.pop());
        e.c[this.i] = o
    }
}
,
G.prototype.v = function(e) {
    for (var t = Buffer.from(e, 'base64').toString('binary'), n = [], r = 0; r < t.length - 1; r += 2)
        n.push(t.charCodeAt(r) << 8 | t.charCodeAt(r + 1));
    this.b = n
}
,
G.prototype.y = function(e) {
    for (var t = Buffer.from(e, 'base64').toString('binary'), n = 66, r = [], o = 0; o < t.length; o++) {
        var i = 24 ^ t.charCodeAt(o) ^ n;
        r.push(String.fromCharCode(i)),
        n = i
    }
    return r.join("")
}
,
G.prototype.F = function(e) {
    var t = this;
    this.g = e.map(function(e) {
        return "string" == typeof e ? t.y(e) : e
    })
}
,
G.prototype.J = function(e, t, n) {
    for (t = t || 0,
    n = n || [],
    this.o = t,
    "string" == typeof e ? (this.F(n),
    this.v(e)) : (this.b = e,
    this.g = n),
    this.G = !0,
    this.x = Date.now(); this.G; ) {
        var r = this.b[this.o++];
        if ("number" != typeof r)
            break;
        var o = Date.now();
        if (500 < o - this.x)
            return;
        this.x = o;
        try {
            this.M(r)
        } catch (e) {
            if (this.R = e,
            !this.Q)
                throw "execption at " + this.o + ": " + e;
            this.o = this.Q
        }
    }
}
,
G.prototype.M = function(e) {
    var t = (61440 & e) >> 12;
    new this.D[t](e).M(this)
}
,
(new G).J("4AeTAJwAqACcAaQAAAAYAJAAnAKoAJwDgAWTACwAnAKoACACGAESOTRHkQAkAbAEIAMYAJwFoAASAzREJAQYBBIBNEVkBnCiGAC0BjRAJAAYBBICNEVkBnDGGAC0BzRAJACwCJAAnAmoAJwKoACcC4ABnAyMBRAAMwZgBnESsA0aADRAkQAkABgCnA6gABoCnA+hQDRHGAKcEKAAMQdgBnFasBEaADRAkQAkABgCnBKgABoCnBOhQDRHZAZxkrAUGgA0QJEAJAAYApwVoABgBnG6sBYaADRAkQAkABgCnBegAGAGceKwGBoANECRACQAnAmoAJwZoABgBnIOsBoaADRAkQAkABgCnBugABoCnByhQDRHZAZyRrAdGgA0QJEAJAAQACAFsB4gBhgAnAWgABIBNEEkBxgHEgA0RmQGdJoQCBoFFAE5gCgFFAQ5hDSCJAgYB5AAGACcH4AFGAEaCDRSEP8xDzMQIAkQCBoFFAE5gCgFFAQ5hDSCkQAkCBgBGgg0UhD/MQ+QACAIGAkaBxQBOYGSABoAnB+EBRoIN1AUCDmRNJMkCRAIGgUUATmAKAUUBDmENIKRACQIGAEaCDRSEP8xD5AAIAgYCRoHFAI5gZIAGgCcH4QFGgg3UBQQOZE0kyQJGAMaCRQ/OY+SABoGnCCEBTTAJAMYAxoJFAY5khI/Nk+RABoGnCCEBTTAJAMYAxoJFAw5khI/Nk+RABoGnCCEBTTAJAMYAxoJFBI5khI/Nk+RABoGnCCEBTTAJAMYBxIDNEEkB3JsHgNQAA==", 0, ["BRgg", "BSITFQkTERw=", "LQYfEhMA", "PxMVFBMZKB8DEjQaBQcZExMC", "", "NhETEQsE", "Whg=", "Wg==", "MhUcHRARDhg=", "NBcPBxYeDQMF", "Lx4ODys+GhMC", "LgM7OwAKDyk6Cg4=", "Mx8SGQUvMQ==", "SA==", "ORoVGCQgERcCAxo=", "BTcAERcCAxo=", "BRg3ABEXAgMaFAo=", "SQ==", "OA8LGBsP", "GC8LGBsP", "Tg==", "PxAcBQ==", "Tw==", "KRsJDgE=", "TA==", "LQofHg4DBwsP", "TQ==", "PhMaNCwZAxoUDQUeGQ==", "PhMaNCwZAxoUDQUeGTU0GQIeBRsYEQ8=", "Qg==", "BWpUGxkfGRsZFxkbGR8ZGxkHGRsZHxkbGRcZG1MbGR8ZGxkXGRFpGxkfGRsZFxkbGR8ZGxkHGRsZHxkbGRcZGw==", "ORMRCyk0Exk8LQ==", "ORMRCyst"]);
var Encryption = function(e) {
    return __g._encrypt(e)
};


/////////////////////////////////////////////


capchaAPI = `https://www.zhihu.com/api/v3/oauth/captcha?lang=en`;
loginAPI = 'https://www.zhihu.com/api/v3/oauth/sign_in'

var captcha = '86TG'

loginData = {
    'client_id': 'c3cef7c66a1843f8b3a9e6a1e3160e20',
    'grant_type': 'password',
    'source': 'com.zhihu.web',
    'username': '+8618324748963',
    'password': 'niudai123.',
    'lang': 'en',
    'ref_source': 'homepage',
	'utm_source': '',
	'captcha': captcha,
	'timestamp': '',
	'signature': ''
}

headers = {
	'accept-encoding': 'gzip',
	'Host': 'www.zhihu.com',
	'Referer': 'https://www.zhihu.com/',
	'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 ' +
	 '(KHTML, like Gecko) Chrome/66.0.3359.181 Safari/537.36',
	 'content-type': 'application/x-www-form-urlencoded',
	 'x-zse-83': '3_1.1',
     'x-xsrftoken': 'dCyt1Kb97IN7jeh5SJo92A9mw2bvv9Es',
}

cookies = fs.readFileSync(path.join(__dirname, 'cookie.txt'), 'utf8')

console.log(cookies);

// getXsrf = function () {
	
// }

headers['cookie'] = cookies;


loginData.timestamp = Math.round(new Date().getTime());
loginData.signature = crypto.createHmac('sha1', 'd1b964811afb40118a12068ff74a12f4')
	.update(loginData.grant_type + loginData.client_id + loginData.source + loginData.timestamp.toString())
	.digest('hex');


console.log(loginData);
console.log(formurlencoded(loginData));
encryptedFormData = Encryption(formurlencoded(loginData));
console.log(encryptedFormData)
console.log(headers)


// captcha = '------WebKitFormBoundarytc1IIDS9ErxdnVS9\r\nContent-Disposition: form-data; name=\"input_text\"\r\n\r\n'
//  + loginData.captcha + 
//  '\r\n------WebKitFormBoundarytc1IIDS9ErxdnVS9--\r\n'

console.log(captcha)

console.log(headers)

var options = {
    method: 'POST',
    uri: capchaAPI,
    form: {
        // Like <input type="text" name="name">
        input_text: captcha
    },
    headers: {
        'cookie': cookies
        /* 'content-type': 'application/x-www-form-urlencoded' */ // Is set automatically
    }
}

httpClient(options).then((body => {
    console.log(body)
    postLoginData()
}))

function postLoginData() {
    httpClient(loginAPI, { method: 'post', headers, body: encryptedFormData, gzip:true }, (error, resp) => {
        let cookieStr = ''
        resp.headers['set-cookie'].forEach(
            c => { 
                c = c.split(';')[0];
                cookieStr = cookieStr.concat(c, '; ') }
        )
        console.log(resp.headers['set-cookie']);
        fs.appendFileSync(path.join(__dirname, 'cookie.txt'), cookieStr, { encoding: 'utf8'})
        console.log(resp.body);
    })    
}
















